design-documents/decrypt-api-token.md

```diff
@@ -0,0 +1,255 @@
+&nbsp;&nbsp;&nbsp;&nbsp;Authentication tokens used to access web APIs (REST, SOAP, GraphQL) should contain enough
+information to authenticate a user without having to store session data on server side.
+ 
+This approach has following advantages:
+* In case of distributed deployment instances don't have to access single storage/service to authenticate a user
+* Saves storage space
+* Faster to decode/decrypt than load from storage
+* Don't have to implement token protection mechanism on backend to keep tokens from being stolen
+ 
+ 
+### Current situation
+&nbsp;&nbsp;&nbsp;&nbsp;Magento stores tokens in RDMS table named `oauth_token` which is originally meant to store long-term integration tokens
+but also used for admin/customer user tokens. Token rows also contain user type, issued data and expiration data which
+is retrieved when authenticating users for web API via token passed in the header. Admin and customer tokens may be
+manually expired by admin users by updating `oauth_token` table data. Tokens themselves are random strings of static
+length.
+ 
+With described above circumstances in order to perform authentication Magento instances have to have access to same DB
+and perform (relatively) expensive SQL queries to DB. Te more users Magento has the more tokens it has to store
+until they expire and run an extra cron job to remove expired ones.
+ 
+ 
+### Overview of new web API tokens implementation
+* `TokenUserContext` to be updated to decode new tokens and use old logic for existing tokens
+* [Create and implement](#spi-for-token-data-authenticator) SPIs for providing user data to be encoded in tokens and reading it
+* [Create and implement](#new-authentication-tokens-api) API for the new tokens generation and reading
+* Update token revoke functionality to use cache storage to store revoke request instead of DB for customer/admin tokens
+  [details](#how-to-handle-manual-expiration)
+ 
+ 
+### New Authentication Tokens API
+&nbsp;&nbsp;&nbsp;&nbsp;There can be many applications possible for the new tokens in Magento: user authentication,
+webhooks, integrations, we need provide an API for developers to establish a mechanism for all future cases.
+ 
+_API:_
+ 
+```php
+/**
+ * Different implementations may add additional data like algorithm, key version etc.
+ */
+interface TokenEncodingDataInterface
+{
+    /**
+     * Key to use when encrypting/signing tokens.
+     *
+     * @return string
+     */
+    public function getKey(): string;
+}
+
+class TokenDecodingException extends \RuntimeException {}
+
+class TokenAuthenticationException extends \InvalidArgumentException {}
+
+interface TokenManagerInterface
+{
+    /**
+     * Create token with encoded authentication data.
+     *
+     * @param array $data Array of scalars, authentication data to be encoded.
+     * @param TokenEncodingDataInterface $data
+     * @return string Token.
+     */
+    public function encode(array $data, TokenEncodingDataInterface $data): string;
```

Comment:
Do you have a prototype for those interfaces? How the implementation of `TokenEncodingDataInterface` will look like for the client code?

Reply:
It's just a DTO which, depending on implementation used will contain secret key, algorithm name, claims list etc